teuthology/task/ceph_ansible.py

```diff
@@ -113,18 +188,157 @@ def generate_hosts_file(self):
+            (remote,) = self.ctx.cluster.only('mon.a').remotes
+            remote.run(args=['sudo', 'ceph', 'osd', 'tree'])
+            remote.run(args=['sudo', 'ceph', '-s'])
+            log.info("Waiting for Ceph health to reach HEALTH_OK \
+                        or HEALTH WARN")
+            while proceed():
+                out = StringIO()
+                remote.run(args=['sudo', 'ceph', 'health'], stdout=out)
+                out = out.getvalue().split(None, 1)[0]
+                log.info("cluster in state: %s", out)
+                if out in ('HEALTH_OK', 'HEALTH_WARN'):
+                    break
+
     def get_host_vars(self, remote):
         extra_vars = self.config.get('vars', dict())
         host_vars = dict()
         if not extra_vars.get('osd_auto_discovery', False):
-            host_vars['devices'] = get_scratch_devices(remote)
+            roles = self.ctx.cluster.remotes[remote]
+            dev_needed = len([role for role in roles
+                              if role.startswith('osd')])
+            host_vars['devices'] = get_scratch_devices(remote)[0:dev_needed]
         if 'monitor_interface' not in extra_vars:
             host_vars['monitor_interface'] = remote.interface
         if 'public_network' not in extra_vars:
             host_vars['public_network'] = remote.cidr
         return host_vars
 
+    def run_rh_playbook(self):
+        ceph_installer = self.ceph_installer
+        args = self.args
+        ceph_installer.run(args=['cp', '-R',
+                                 '/usr/share/ceph-ansible', '.'])
+        ceph_installer.put_file(self.inventory, 'ceph-ansible/inven.yml')
+        ceph_installer.put_file(self.playbook_file, 'ceph-ansible/site.yml')
+        # copy extra vars to groups/all
+        ceph_installer.put_file(self.gvars, 'ceph-ansible/group_vars/all')
+        # print for debug info
+        ceph_installer.run(args=('cat', 'ceph-ansible/inven.yml'))
+        ceph_installer.run(args=('cat', 'ceph-ansible/site.yml'))
+        ceph_installer.run(args=('cat', 'ceph-ansible/group_vars/all'))
+        out = StringIO()
+        str_args = ' '.join(args)
+        ceph_installer.run(args=['cd', 'ceph-ansible', run.Raw(';'),
+                                 run.Raw(str_args)],
+                           timeout=4200,
+                           check_status=False,
+                           stdout=out)
+        log.info(out.getvalue())
+        if re.search(r'all hosts have already failed', out.getvalue()):
+            log.error("Failed during ceph-ansible execution")
+            raise CephAnsibleError("Failed during ceph-ansible execution")
+
+    def run_playbook(self):
+        # setup ansible on first mon node
+        ceph_installer = self.ceph_installer
+        args = self.args
+        if ceph_installer.os.package_type == 'rpm':
+            # install crypto packages for ansible
+            ceph_installer.run(args=['sudo',
```

Comment:
These blocks would be much nicer formatted like:

```
ceph_installer.run(args=[
    'sudo',
    'yum',
    'install',
    'ceph',
])
```

This has the advantage of producing nicer diffs if this code changes and it's easier to read, IMHO.

I wouldn't block on this change, but it'd be very nice to have.